### PR TITLE
Integrate FSDP2 wrapper into Trainer

### PR DESCRIPTION
**Pull request by bowenyang008**
# What does this PR do?
As titled, including:
1. Support construct FSDP2Config from legacy/FSDP1 args
2. Depending on FSDP(2)Config or env FSDP_VERSION, auto construct FSDP1 or FSDP2 directly in Trainer
3. Refactor FSDP2 utils to work for ComposerModel and backward compatible with FSDP1 on:
    * auto wrap
    * activation checkpointing
    * meta init
4. Update unit tests to adapt to 2 & 3

# Testing
1. All unit tests are updated
2. Additional unit test added for FSDP2Config
3. Manual e2e tests on MPT 1B/7B models, see below
   * Accuracy matches (with same init weights)
   * Active Memory: 2-7% better
   * Throughput: 3% gap in 7B, 30% in 1B
4. More unit tests should be added for in following PRs:
   * GradNorm
   * Meta Init
5. Need e2e regression test

Loss/Acc:
1B:
loss_train_total
metrics_train_TokenAccuracy
7B:
loss_train_total
metrics_train_TokenAccuracy

MFU:
1B:
throughput_device_mfu
7B:
throughput_device_mfu

Peak active Memory:
1B:
memory_peak_active_mem
7B:
memory_peak_active_mem

Peak Reserved Memory:
1B:
memory_peak_reserved_mem
7B:
memory_peak_reserved_mem

### CODE REVIEW

File: composer/distributed/prepare_distributed.py [hunk 1]
```diff
@@ -3,40 +3,64 @@
...
+        param_init_fn (Callable[[torch.nn.Module], None]): Function to initialize model parameters
+            after FSDP wrapping. Defaults to a no-op function.
+
+    Raises:
+        ValueError: If the config is not an FSDP2Config or if activation_checkpointing_check_fn is provided
+            but neither activation_checkpointing nor activation_cpu_offload is enabled in the config.
     """
-    if isinstance(config, FSDPConfig):
-        raise ValueError('FSDPConfig is not supported for now, use FSDP2Config instead')
+    if not isinstance(config, FSDP2Config):
```
Comment (dakinggg) [inline composer/distributed/prepare_distributed.py:51]: can just delete this error - the typing is sufficient

File: composer/distributed/prepare_distributed.py [hunk 2]
```diff
@@ -3,40 +3,64 @@
...
+        raise ValueError('FSDP2Config is the only supported config for now')
 
     if activation_checkpointing_check_fn is not None:
         if not config.activation_checkpointing and not config.activation_cpu_offload:
             raise ValueError(
                 'Activation checkpointing or offloading must be enabled if activation_checkpointing_check_fn is provided',
             )
+    # Use the context manager for optimizer synchronization if optimizer is provided
+    with sync_optimizer_and_model_params(optimizer, model) if optimizer is not None else nullcontext():
```
Comment (dakinggg) [inline composer/distributed/prepare_distributed.py:60]: noting that this change switches the order of apply_ac and prepare_fully_shared. Is that ok? Either order works?

File: composer/distributed/prepare_distributed.py [hunk 3]
```diff
@@ -46,4 +70,34 @@ def parallelize_model(
             activation_checkpointing_check_fn,
         )
 
-    prepare_fully_shard(model, optimizer, config, fsdp_wrap_policy)
+
+def parallelize_composer_model(
```
Comment (dakinggg) [inline composer/distributed/prepare_distributed.py:74]: this function is called by trainer in the fsdp2 case, but uses functions like `generate_fsdp1...`. Whats the logic behind the naming of those?

**Comment by dakinggg** [comment]
@bowenyang008 can you show the wrapping of the model that occurs with each of fsdp1 and fsdp2 for your manual test runs? in fsdp1 config you can set `verbose` to True, may need to make an equivalent for fsdp2. want to be able to see that they are wrapped the same.

File: composer/algorithms/gradient_clipping/gradient_clipping.py [hunk 4]
```diff
@@ -137,7 +137,7 @@ def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
                 model=state.model,
                 clipping_type=self.clipping_type,
                 clipping_threshold=self.clipping_threshold,
-                fsdp_enabled=state.fsdp_enabled,
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/algorithms/gradient_clipping/gradient_clipping.py:140]: yup, [torchtitan](https://github.com/pytorch/torchtitan/blob/14c810ed08af9f76070ad970b05ee3f04d0e9e15/docs/fsdp.md) mentioned it, and apparently I forgot what I read 🤦

File: tests/common/models.py [hunk 5]
```diff
@@ -221,6 +221,8 @@ def __init__(self, num_features: int = 1, device: str = 'cpu') -> None:
         self.mlp.fc1.weight = self.mlp.fc2.weight
 
     def add_fsdp_wrap_attribute_to_children(self):
+        for child in self.children():
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply tests/common/models.py:224]: because some of the test are testing https://github.com/mosaicml/composer/blob/boweny/fsdp2/trainer-integration/composer/distributed/prepare_distributed.py#L20 directly rather than https://github.com/mosaicml/composer/blob/boweny/fsdp2/trainer-integration/composer/distributed/prepare_distributed.py#L74. The former does not have a default policy that mimics FSDP1, the latter does

File: composer/utils/parallelism.py [hunk 6]
```diff
@@ -81,6 +81,45 @@ class FSDP2Config:
     activation_checkpointing: bool = False
     activation_cpu_offload: bool = False
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/utils/parallelism.py:84]: I don't have an intention to support inheritance of FSDP2Config, as the next version can be quite different than current version (e.g., the Simple FSDP in torchtitan) just like FSDP2 is quite different from FSDP1

File: composer/core/state.py [hunk 7]
```diff
@@ -1694,3 +1715,31 @@ def precision_config(self):
     def is_model_ddp(self):
         """Whether :attr:`model` is an instance of a :class:`.DistributedDataParallel`."""
         return isinstance(self.model, DistributedDataParallel)
+
+    def log_post_backward_param_and_gradient(self):
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/core/state.py:1719]: sure, I plan to work on the TODO shortly in a follow up PR so left this code piece temporarily, I can remove it for now

File: composer/distributed/activation_checkpointing.py [hunk 8]
```diff
@@ -31,6 +32,41 @@ def _check_fn(module: torch.nn.Module) -> bool:
...
+
+    Args:
+        composer_model (nn.Module): The ComposerModel to generate a check function for.
+
+    Returns:
+        Callable: A function that determines whether a module's activations should be checkpointed.
+    """
+    cached_submodules_ac: dict[nn.Module, bool] = {composer_model: False}
+    for child in composer_model.children():
+        cached_submodules_ac[child] = True
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/activation_checkpointing.py:51]: basically mimics FSDP1 logic, all direct children of root module are ACed

File: composer/distributed/fsdp2.py [hunk 9]
```diff
@@ -82,7 +88,14 @@ def apply_fully_shard(
...
     """
     # Define the default kwargs for fully_shard
-    fully_shard_kwargs = {'mesh': fsdp2_config.device_mesh, 'reshard_after_forward': fsdp2_config.reshard_after_forward}
+    # NOTE Model in LLM Foundry mostly uses PURE for its MP policy so we default to both bfloat16 for now
+    # yet both Composer and TorchTitan's default mp_policy use bfloat16 for params all-gather and float32 for reduce-scatter
+    # TODO: support user specified mp_policy
+    fully_shard_kwargs = {
+        'mesh': fsdp2_config.device_mesh,
+        'reshard_after_forward': fsdp2_config.reshard_after_forward,
+        'mp_policy': MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.bfloat16),
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2.py:97]: I thought we agreed we would default to PURE for FSDP2, if it is concerning to you I can prioritize adding the TODO after this PR so the PR don't become too bulky

File: composer/distributed/fsdp2_utils.py [hunk 10]
```diff
@@ -260,30 +303,78 @@ def generate_default_policy(parent_model: nn.Module) -> CustomPolicy:
...
+        composer_model (ComposerModel): The ComposerModel to generate a policy for.
+
+    Returns:
+        CustomPolicy: A policy function that determines which modules to wrap with FSDP.
+
+    Raises:
+        KeyError: If a module's fsdp_wrap_fn returns a dict with invalid FSDP2Config keys.
+    """
+    # Filter the specific deprecation warning to only appear once.
+    warnings.filterwarnings(
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2_utils.py:344]: yup, given line 360 below

File: composer/distributed/prepare_distributed.py (hunk 1 above)
Comment (bowenyang008) [reply composer/distributed/prepare_distributed.py:51]: sure

File: composer/distributed/prepare_distributed.py (hunk 2 above)
Comment (bowenyang008) [reply composer/distributed/prepare_distributed.py:60]: yup, either order works, at least functionally. Theoretically the performance should be the same as well. There is one edge case I would like to test out when I get time though. The reason I have to change its order was it changed the model modules and invalidates the provided policy that only captures pre-dist submodules. I can probably fix this if we see a performance gain of the original order.

File: composer/distributed/prepare_distributed.py (hunk 3 above)
Comment (bowenyang008) [reply composer/distributed/prepare_distributed.py:74]: In FSDP1 wrapper, two things are a bit coupled, e.g., the logic of treating each child module as an independent module and FSDP/AC wrap_fn on each of them. There was a reason for it due to impl of FSDP1 yet that reason is gone in FSDP2. So we have two level of APIs, the main one is the parallelize_model that can take a flexible FSDP or AC wrap policy w/o assuming anything about children module and is similar to torchtitan's parallelize_model for future extensibility, and the second one is the `parallelize_composer_model` that can provide a policy that works with existing fsdp_wrap_fn defined on LLM-Foundry models and produces same FSDP/AC structure as FSDP1

File: composer/distributed/activation_checkpointing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/distributed/activation_checkpointing.py:51]: where do you see that? I thought AC was purely determined by the attr/fn, nothing defaults to True

File: composer/distributed/fsdp2.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/distributed/fsdp2.py:97]: Yes, but PURE does not hardcode to bfloat16, it selects based on the overall precision used for training (https://github.com/mosaicml/composer/blob/a18175f250a806d078b414cda88f51e4275811c9/composer/distributed/mosaic_parallelism.py#L82-L85)

**Comment by bowenyang008** [comment]
> @bowenyang008 can you show the wrapping of the model that occurs with each of fsdp1 and fsdp2 for your manual test runs? in fsdp1 config you can set `verbose` to True, may need to make an equivalent for fsdp2. want to be able to see that they are wrapped the same.

sure let me add a TODO for verbose print, but here is a quick reference:
FSDP1:
```
ComposerMPTCausalLM(
  (model): FullyShardedDataParallel(
    (_fsdp_wrapped_module): MPTForCausalLM(
      (transformer): MPTModel(
        (wte): FullyShardedDataParallel(
          (_fsdp_wrapped_module): SharedEmbedding(50368, 768)
        )
        (wpe): FullyShardedDataParallel(
          (_fsdp_wrapped_module): Embedding(2048, 768)
        )
        (emb_drop): FullyShardedDataParallel(
          (_fsdp_wrapped_module): Dropout(p=0.0, inplace=False)
        )
        (blocks): ModuleList(
          (0-11): 12 x FullyShardedDataParallel(
            (_fsdp_wrapped_module): MPTBlock(
              (norm_1): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
              (attn): MultiheadAttention(
                (Wqkv): Linear(in_features=768, out_features=2304, bias=True)
                (out_proj): Linear(in_features=768, out_features=768, bias=True)
              )
              (norm_2): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
              (ffn): MPTMLP(
                (up_proj): Linear(in_features=768, out_features=3072, bias=True)
                (down_proj): Linear(in_features=3072, out_features=768, bias=True)
              )
              (resid_attn_dropout): Dropout(p=0.0, inplace=False)
              (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
            )
          )
        )
        (norm_f): FullyShardedDataParallel(
          (_fsdp_wrapped_module): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
        )
      )
    )
  )
  (loss_fn): FullyShardedDataParallel(
    (_fsdp_wrapped_module): CrossEntropyLoss()
  )
)
```
FSDP2:
```
ComposerMPTCausalLM(
  (model): FSDPMPTForCausalLM(
    (transformer): MPTModel(
      (wte): FSDPSharedEmbedding(50368, 768)
      (wpe): FSDPEmbedding(2048, 768)
      (emb_drop): FSDPDropout(p=0.0, inplace=False)
      (blocks): ModuleList(
        (0-11): 12 x FSDPMPTBlock(
          (norm_1): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
          (attn): MultiheadAttention(
            (Wqkv): Linear(in_features=768, out_features=2304, bias=True)
            (out_proj): Linear(in_features=768, out_features=768, bias=True)
          )
          (norm_2): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
          (ffn): MPTMLP(
            (up_proj): Linear(in_features=768, out_features=3072, bias=True)
            (down_proj): Linear(in_features=3072, out_features=768, bias=True)
          )
          (resid_attn_dropout): Dropout(p=0.0, inplace=False)
          (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
        )
      )
      (norm_f): FSDPLPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
    )
  )
  (loss_fn): FSDPCrossEntropyLoss()
)
```

File: composer/distributed/prepare_distributed.py (hunk 3 above)
Comment (dakinggg) [reply composer/distributed/prepare_distributed.py:74]: yeah, i think i get that part, but im wondering why you named the functions "fsdp1" when they are being used with our fsdp2 implementation. Why not just call them `generate_fsdp_composer_model_policy` etc

File: composer/distributed/fsdp2.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2.py:97]: oh, I see what you meant, if you don't mind, I will still prefer to support this in a follow up PR to keep things a bit simpler

File: composer/distributed/activation_checkpointing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/activation_checkpointing.py:51]: okay, good catch, I apparently missed that, let me fix it and relaunch a new run

File: composer/distributed/fsdp2.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/fsdp2.py:97]: sg, will likely follow up with a PR before running anything new

File: composer/distributed/prepare_distributed.py (hunk 3 above)
Comment (bowenyang008) [reply composer/distributed/prepare_distributed.py:71]: because the policy is tied to FSDP1 impl. e.g., there is no way to FSDP the root module with FSDP1. But I am okay to change the name to generate_fsdp_composer_model_policy if you prefer

Comment (dakinggg) [reply composer/distributed/prepare_distributed.py:71]: I see, yeah would prefer name change cause rn it looks like a mistake calling fsdp1 code within fsdp2 implementation

**Comment by bowenyang008** [comment]
FSDP + AC:
FSDP1:
```
ComposerMPTCausalLM(
  (model): FullyShardedDataParallel(
    (_fsdp_wrapped_module): MPTForCausalLM(
      (transformer): MPTModel(
        (wte): FullyShardedDataParallel(
          (_fsdp_wrapped_module): SharedEmbedding(50368, 4096)
        )
        (wpe): FullyShardedDataParallel(
          (_fsdp_wrapped_module): Embedding(2048, 4096)
        )
        (emb_drop): FullyShardedDataParallel(
          (_fsdp_wrapped_module): Dropout(p=0.0, inplace=False)
        )
        (blocks): ModuleList(
          (0-1): 2 x FullyShardedDataParallel(
            (_fsdp_wrapped_module): CheckpointWrapper(
              (_checkpoint_wrapped_module): MPTBlock(
                (norm_1): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
                (attn): MultiheadAttention(
                  (Wqkv): Linear(in_features=4096, out_features=12288, bias=True)
                  (out_proj): Linear(in_features=4096, out_features=4096, bias=True)
                )
                (norm_2): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
                (ffn): MPTMLP(
                  (up_proj): Linear(in_features=4096, out_features=16384, bias=True)
                  (down_proj): Linear(in_features=16384, out_features=4096, bias=True)
                )
                (resid_attn_dropout): Dropout(p=0.0, inplace=False)
                (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
              )
            )
          )
        )
        (norm_f): FullyShardedDataParallel(
          (_fsdp_wrapped_module): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
        )
      )
    )
  )
  (loss_fn): FullyShardedDataParallel(
    (_fsdp_wrapped_module): CrossEntropyLoss()
  )
)
```
FSDP2:
```
ComposerMPTCausalLM(
  (model): FSDPMPTForCausalLM(
    (transformer): MPTModel(
      (wte): FSDPSharedEmbedding(50368, 4096)
      (wpe): FSDPEmbedding(2048, 4096)
      (emb_drop): FSDPDropout(p=0.0, inplace=False)
      (blocks): ModuleList(
        (0-1): 2 x CheckpointWrapper(
          (_checkpoint_wrapped_module): FSDPMPTBlock(
            (norm_1): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
            (attn): MultiheadAttention(
              (Wqkv): Linear(in_features=4096, out_features=12288, bias=True)
              (out_proj): Linear(in_features=4096, out_features=4096, bias=True)
            )
            (norm_2): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
            (ffn): MPTMLP(
              (up_proj): Linear(in_features=4096, out_features=16384, bias=True)
              (down_proj): Linear(in_features=16384, out_features=4096, bias=True)
            )
            (resid_attn_dropout): Dropout(p=0.0, inplace=False)
            (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
          )
        )
      )
      (norm_f): FSDPLPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
    )
  )
  (loss_fn): CheckpointWrapper(
    (_checkpoint_wrapped_module): FSDPCrossEntropyLoss()
  )
)
```

File: composer/distributed/activation_checkpointing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bowenyang008) [reply composer/distributed/activation_checkpointing.py]: result updated in comment @dakinggg, since MPTForCausalLM has the fn that set itself to False, so the only diff is the loss_fn. But good catch. Metric wise it is the same before the fix

**Comment by bowenyang008** [comment]
FSDP wrapping AC:
```
ComposerMPTCausalLM(
  (model): FSDPMPTForCausalLM(
    (transformer): MPTModel(
      (wte): FSDPSharedEmbedding(50368, 4096)
      (wpe): FSDPEmbedding(2048, 4096)
      (emb_drop): FSDPDropout(p=0.0, inplace=False)
      (blocks): ModuleList(
        (0-31): 32 x FSDPCheckpointWrapper(
          (_checkpoint_wrapped_module): MPTBlock(
            (norm_1): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
            (attn): MultiheadAttention(
              (Wqkv): Linear(in_features=4096, out_features=12288, bias=True)
              (out_proj): Linear(in_features=4096, out_features=4096, bias=True)
            )
            (norm_2): LPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
            (ffn): MPTMLP(
              (up_proj): Linear(in_features=4096, out_features=16384, bias=True)
              (down_proj): Linear(in_features=16384, out_features=4096, bias=True)
            )
            (resid_attn_dropout): Dropout(p=0.0, inplace=False)
            (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
          )
        )
      )
      (norm_f): FSDPLPLayerNorm((4096,), eps=1e-05, elementwise_affine=True)
    )
  )
  (loss_fn): FSDPCrossEntropyLoss()
)
```
MFU:
throughput_device_mfu

File: composer/distributed/prepare_distributed.py (hunk 2 above)
Comment (bowenyang008) [reply composer/distributed/prepare_distributed.py]: @dakinggg actually I just tested out FSDP post AC, and there is no meaningful diff, see comments

File: composer/utils/parallelism.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/utils/parallelism.py:86]: yeah, im not really expecting inheritance, but still think its better to use cls for these. not a big deal either way, can ignore

**Review by dakinggg — APPROVED** [review]
lgtm, should get ricky to approve also

File: composer/distributed/fsdp2.py [hunk 11]
```diff
@@ -7,34 +7,38 @@
...
     module: nn.Module,
+    visited_modules: set[nn.Module],
     target_modules_to_kwargs: dict[nn.Module, dict],
 ) -> None:
     """Recursive helper to apply fully_shard based on policy and legalization.
 
     Args:
         root_module (nn.Module): The root module to check for parameter sharing.
         module (nn.Module): The current module being processed.
+        visited_modules (set[nn.Module]): Set of modules that have already been visited so we don't process them twice.
```
Comment (rithwik-db) [inline composer/distributed/fsdp2.py:33]: Oh curious, did you see an issue where we were processing nodes more times than necessary?

Comment (bowenyang008) [reply composer/distributed/fsdp2.py:33]: yup, just a simple case M1(M2, M3(M2)) would run into this case if we fsdp both M2 and M3, not sure why we didn't run into this before

File: composer/distributed/fsdp2_utils.py [hunk 12]
```diff
@@ -260,30 +303,78 @@ def generate_default_policy(parent_model: nn.Module) -> CustomPolicy:
...
+        fsdp_wrap_fn = getattr(child, 'fsdp_wrap_fn', lambda x: cached_submodules_to_wrap.get(x, False))
+        for module in child.modules():
+            if hasattr(module, '_fsdp_wrap'):
+                warnings.warn(
+                    DeprecationWarning(
+                        'The _fsdp_wrap attribute will be removed in a future release. Please use fsdp_wrap_fn instead.',
+                    ),
+                )
+                cached_submodules_to_wrap[module] = bool(module._fsdp_wrap)
+            elif module == child:
```
Comment (rithwik-db) [inline composer/distributed/fsdp2_utils.py:367]: nit: would this be better off as "child_module" instead of "module" and "child_root" instead of "child"? I think it would make this a bit more readable. I'm not entirely opposed to keeping it as is though

**Review by rithwik-db — APPROVED** [review]
Looks good, thanks!
